libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/MapboxTripSessionNoSetupTest.kt

```diff
@@ -0,0 +1,331 @@
+            .takeLast(2) // take only events triggered by location updates
+            .map { it.voiceInstructions?.announcement() }
+        assertEquals(listOf("1", "2"), voiceInstructionsAnnouncements)
+    }
+
+    @Test
+    fun newVoiceInstructionsTriggerEvenIfTheyAreTheSameAsPreviousOne() {
+        // arrange
+        val nativeNavigator = mockk<MapboxNativeNavigator>(relaxed = true)
+        StatusWithVoiceInstructionUpdateUtil.triggerStatusUpdatesOnLocationUpdate(nativeNavigator)
+        val locationEngine = TestLocationEngine.create()
+        val tripSession = buildTripSession(
+            nativeNavigator = nativeNavigator,
+            locationEngine = locationEngine
+        )
+        val voiceInstructionObserver = VoiceInstructionsObserverRecorder()
+        tripSession.registerVoiceInstructionsObserver(voiceInstructionObserver)
+        tripSession.start(true)
+        tripSession.setRoutes(
+            listOf(createDirectionsRoute()),
+            0,
+            ROUTES_UPDATE_REASON_NEW
+        )
+        locationEngine.updateLocation(createLocation(longitude = LONGITUDE_FOR_VOICE_INSTRUCTION_1))
+        // act
+        locationEngine.updateLocation(createLocation(longitude = LONGITUDE_FOR_VOICE_INSTRUCTION_1))
+        // assert
+        val newVoiceInstructions = voiceInstructionObserver.records
+            .takeLast(2) // take only events triggered by location updates
+            .map { it.announcement() }
+        assertEquals(listOf("1", "1"), newVoiceInstructions)
+    }
+}
+
+private fun buildTripSession(
+    nativeNavigator: MapboxNativeNavigator = mockk(relaxed = true),
+    locationEngine: LocationEngine = TestLocationEngine.create()
+): MapboxTripSession {
+    val context: Context = ApplicationProvider.getApplicationContext()
+    val navigationOptions = NavigationOptions.Builder(context)
+        .locationEngine(locationEngine)
+        .build()
+
+    val tripService: TripService = mockk(relaxUnitFun = true) {
+        every { hasServiceStarted() } returns false
+    }
+
+    val parentJob = SupervisorJob()
+    val testScope = CoroutineScope(parentJob + TestCoroutineDispatcher())
+    val threadController = spyk<ThreadController>()
+    every { threadController.getMainScopeAndRootJob() } returns JobControl(parentJob, testScope)
+
+    return MapboxTripSession(
+        tripService,
+        TripSessionLocationEngine(navigationOptions),
+        nativeNavigator,
+        threadController,
+        logger = mockk(relaxed = true),
+        eHorizonSubscriptionManager = mockk(relaxed = true),
+    )
+}
+
+class StatusWithVoiceInstructionUpdateUtil {
+    companion object {
+
+        const val LONGITUDE_FOR_VOICE_INSTRUCTION_1 = 1.0
+        const val LONGITUDE_FOR_VOICE_INSTRUCTION_NULL = 1.5
+        const val LONGITUDE_FOR_VOICE_INSTRUCTION_2 = 2.0
+
+        fun triggerStatusUpdatesOnLocationUpdate(nativeNavigator: MapboxNativeNavigator) {
```

Comment:
nit: make `StatusWithVoiceInstructionUpdateUtil` object instead class and restrict companion object

Reply:
Can't explain why, but I like that `triggerStatusUpdatesOnLocationUpdate` becomes static function in the bytecode. But you find it very ugly, I'm ready to accept singleton in the bytecode and replace `companion` by `object`. 🙂

Reply:
It doesn't become a static function in the bytecode, unless you add a `JvmStatic` annotation. 
It is actually a singleton `StatusWithVoiceInstructionUpdateUtil.Companion`. 
So basically there is no difference, but with `object` it is simpler. 
Also I think `triggerStatusUpdatesOnLocationUpdate` can just be a private function in test class or file.

Reply:
`StatusWithVoiceInstructionUpdateUtil.Companion` is actually a static class in bytecode. There're 2 classes, and we can restrict it to only one

Reply:
Ok, it's not how it works. I will fix

Reply:
fixed, thank you guys 👍